File: Modules/PathPlannings/SCurve/src/s_curve.cpp
/**
 * @file    s_curve.cpp
 * @author  syhanjin
 * @date    2026-01-28
 */
#include "s_curve.hpp"

#include <cmath>

SCurve::SCurveAccel::SCurveAccel() :
    has_uniform_(false), vs_(0), jm_(0), total_time_(0), total_distance_(0), t1_(0), x1_(0), v1_(0),
    t2_(0), x2_(0), v2_(0), ap_(0), vp_(0)
{
}

void SCurve::SCurveAccel::init(const float vs, const float vp, const float am, const float jm)
{
    has_uniform_ = jm * (vp - vs) > am * am;
    vs_          = vs;
    vp_          = vp;
    jm_          = jm;
    if (has_uniform_)
    {
        ap_ = am;

        t1_             = am / jm;
        t2_             = (vp - vs) / am;
        const float dt2 = t2_ - t1_;

        v1_ = vs + 0.5f * am * t1_;
        v2_ = vp - 0.5f * am * t1_;

        x1_ = vs * t1_ + 1 / 6.0f * am * t1_ * t1_;
        x2_ = x1_ + v1_ * dt2 + 0.5f * am * dt2 * dt2;

        total_time_     = t2_ + t1_;
        total_distance_ = (vp * vp - vs * vs) / (2.0f * am) +
                          0.5f * (vs + vp) * t1_ /*(vs + vp) * am / (2.0f * jm)*/;
    }
    else
    {
        ap_ = std::sqrtf(jm * (vp - vs));

        t1_ = ap_ / jm;
        t2_ = t1_; // 没有匀加速段

        v1_ = vs + 0.5f * ap_ * ap_ / jm;
        v2_ = v1_;

        x1_ = vs * t1_ + 1 / 6.0f * ap_ * t1_ * t1_;
        x2_ = x1_;

        total_time_     = 2.0f * sqrtf((vp - vs) / jm);
        total_distance_ = (vs + vp) * sqrtf((vp - vs) / jm);
    }
}
float SCurve::SCurveAccel::getDistance(const float t) const
{
    if (t <= 0)
    {
        return 0;
    }
    if (t < t1_)
    {
        return vs_ * t + 1 / 6.0f * jm_ * t * t * t;
    }
    // 除了这段以外两种情况都一样
    if (has_uniform_ && t < t2_)
    {
        const float _t = t - t1_;
        return x1_ + v1_ * _t + 0.5f * ap_ * _t * _t;
    }
    if (t < total_time_)
    {
        const float _t = total_time_ - t;
        return total_distance_ - vp_ * _t + 1 / 6.0f * jm_ * _t * _t * _t;
    }
    return total_distance_;
}

float SCurve::SCurveAccel::getVelocity(const float t) const
{
    if (t <= 0)
    {
        return vs_;
    }
    if (t < t1_)
    {
        return vs_ + 0.5f * jm_ * t * t;
    }
    if (has_uniform_ && t < t2_)
    {
        return v1_ + ap_ * (t - t1_);
    }
    if (t < total_time_)
    {
        const float _t = total_time_ - t;
        return vp_ - 0.5f * jm_ * _t * _t;
    }
    return vp_;
}

float SCurve::SCurveAccel::getAcceleration(const float t) const
{
    if (t <= 0)
        return 0;
    if (t < t1_)
        return jm_ * t;
    if (has_uniform_ && t < t2_)
        return ap_;
    if (t < total_time_)
        return jm_ * (total_time_ - t);
    return 0;
}

SCurve::SCurve() :
    has_const_(false), direction_(0), vp_(0), vs_(0), as_(0), jm_(0), t0_(0), x0_(0), xs_(0),
    x1_(0), x2_(0), xe_(0), ts1_(0), xs1_(0), t1_(0), t2_(0), total_time_(0)
{
#ifdef DEBUG
    binary_search_count_ = 0;
#endif
}

SCurve::Result SCurve::init(
        const float xs, const float xe, float vs, float as, float vm, float am, float jm)
{
    vm = fabsf(vm), am = fabsf(am), jm = fabsf(jm);

    // 全部归到正向移动判断
    const float dir = xe > xs ? 1.0f : -1.0f;
    direction_      = dir;
    xs_             = xs;
    xe_             = xe;
    const float len = fabsf(xe - xs);
    vs = vs * dir, as = as * dir;
    vs_ = vs, as_ = as;
    jm_ = jm;

    // 如果目标基本没有变化，则不动
    if (len < 1e-6f)
    {
        t0_         = 0;
        t1_         = 0;
        has_const_  = false;
        t2_         = 0;
        total_time_ = 0;
    }

    if (fabsf(vs) > vm || fabsf(as) > am)
        // 如果初速度或者初加速度超限，则生成失败
        return Result::Failed;

    // 首先假定存在匀速段
    const float vp = vm;
    float       vs0, dx0, vp_min;

    float vs1, ts1;

    if (as < 0)
    {
        // 如果初加速度与目标方向相反，则必须先刹到 0
        // 这段位移是必须的

        // 加速度刹到 0 之后的速度，作为之后过程的初速度
        vs0 = vs - 0.5f * as * as / jm;
        // 已经超限则无法构建曲线
        if (fabsf(vs0) > vm)
            return Result::Failed;
        vp_min = vs0;

        const float ts0 = -as / jm;
        t0_             = ts0;

        // 加速度刹到 0 的位移
        dx0 = vs * ts0 + 1 / 3.0f * as * ts0 * ts0;
        x0_ = xs + dir * dx0;
        vs1 = vs0; // 之后无须构造偏移
        ts1 = 0;
    }
    else
    {
        // 如果初始带有同方向的加速度，必须预留速度刹车
        vp_min = vs + 0.5f * as * as / jm;
        vs0    = vs;
        dx0    = 0;
        t0_    = 0;
        x0_    = xs;
        // 构造带偏移的梯形加速
        if (vm < vp_min)
            // 即使 vp = vm 也无法构造出带偏移的梯形加速
            return Result::Failed;
        ts1 = as / jm;
        vs1 = vs - 0.5f * as * ts1;
    }
    if (vp_min < 0)
        vp_min = 0;
    // 实际需要使用的 len0
    const float len0 = len - dx0;
    {
        process1_.init(vs1, vp, am, jm);
        // 构造逆过程梯形加速
        process3_.init(0, vp, am, jm);
        xs1_                = process1_.getDistance(ts1);
        const float dx1     = process1_.getTotalDistance() - xs1_;
        const float dx3     = process3_.getTotalDistance();
        const float x_const = len0 - dx1 - dx3;
        if (x_const > 0)
        {
            // 存在匀速段
            has_const_ = true;
            ts1_       = ts1;

            // 这里是时刻分界点，所以需要加上开头部分
            t1_ = t0_ + process1_.getTotalTime() - ts1;

            const float t_const = x_const / vm;
            t2_                 = t1_ + t_const;
            total_time_         = t2_ + process3_.getTotalTime();

            x1_ = x0_ + direction_ * dx1;
            x2_ = x1_ + direction_ * x_const;

            vp_ = vm;
            return Result::Success;
        }
    }
    // 不存在匀速段，求最大速度
    // 由于代数表达式太过复杂，这里直接上二分查找
    float l = vp_min, r = vm;
    float delta_d = len0, dx1 = 0, dx3 = 0;

#ifdef DEBUG
    binary_search_count_ = 0;
#endif
    // 最大迭代次数约 13 次
    while (r - vp_min > 0.001f)
    {
#ifdef DEBUG
        binary_search_count_++;
#endif
        const float mid = 0.5f * (l + r);
        process1_.init(vs1, mid, am, jm);
        process3_.init(0, mid, am, jm);
        // 更新，用于判断是否找到解
        dx1     = process1_.getTotalDistance() - process1_.getDistance(ts1);
        dx3     = process3_.getTotalDistance();
        delta_d = dx1 + dx3 - len0;
        if (delta_d < S_CURVE_MAX_BS_ERROR && delta_d > -S_CURVE_MAX_BS_ERROR)
        {
            r = l = mid;
            break;
        }
        if (delta_d > 0)
            r = mid;
        else
            l = mid;
    }
    if (delta_d > S_CURVE_MAX_BS_ERROR)
    {
        // 即使 vp 降到最低也无法找到解
        return Result::Failed;
    }

    has_const_ = false;
    ts1_       = ts1;
    // 这里是时刻分界点，所以需要加上开头部分
    t1_         = t0_ + process1_.getTotalTime() - ts1;
    t2_         = t1_;
    total_time_ = t2_ + process3_.getTotalTime();

    x1_ = x0_ + direction_ * dx1;
    x2_ = x1_;

    vp_ = r;
    return Result::Success;
}

float SCurve::CalcX(const float t) const
{
    // 起始之前
    if (t <= 0)
        return xs_;
    // 反向加速度刹车
    if (t < t0_)
    {
        const float t2 = t * t;
        const float t3 = t2 * t;
        return xs_ + direction_ * (vs_ * t + 0.5f * as_ * t2 + 1 / 6.0f * jm_ * t3);
    }
    // 加速过程
    if (t < t1_)
        return x0_ + direction_ * (process1_.getDistance(t - t0_ + ts1_) - xs1_);
    // 匀速过程
    if (has_const_ && t < t2_)
        return x1_ + direction_ * vp_ * (t - t1_);
    // 减速过程
    if (t < total_time_)
        return xe_ - direction_ * process3_.getDistance(total_time_ - t);
    return xe_;
}

float SCurve::CalcV(const float t) const
{
    // 起始之前
    if (t <= 0)
        return direction_ * vs_;
    // 反向加速度刹车
    if (t < t0_)
        return direction_ * (vs_ + as_ * t + 0.5f * jm_ * t * t);
    // 加速过程
    if (t < t1_)
        return direction_ * process1_.getVelocity(t - t0_ + ts1_);
    // 匀速过程
    if (has_const_ && t < t2_)
        return direction_ * vp_;
    // 减速过程
    if (t < total_time_)
        return direction_ * process3_.getVelocity(total_time_ - t);
    return 0;
}

float SCurve::CalcA(const float t) const
{
    // 起始之前
    if (t <= 0)
        return direction_ * as_;
    // 反向加速度刹车
    if (t < t0_)
        return direction_ * (as_ + jm_ * t);
    // 加速过程
    if (t < t1_)
        return direction_ * process1_.getAcceleration(t - t0_ + ts1_);
    // 匀速过程
    if (has_const_ && t < t2_)
        return 0;
    // 减速过程
    if (t < total_time_)
        return -direction_ * process3_.getAcceleration(total_time_ - t);
    return 0;
}

File: Modules/PathPlannings/VelocityProfile/SCurve/simulation/s_curve.py
# -*- coding: utf-8 -*-

"""
本文件是通过 AI 将 s_curve.c 转换得来
--------------------------------------------------------------------------
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
import math


class SCurveAccel:
    def __init__(self):
        self.has_uniform = False
        self.vs = 0.0
        self.jm = 0.0

        self.total_time = 0.0
        self.total_distance = 0.0

        self.t1 = 0.0
        self.x1 = 0.0
        self.v1 = 0.0
        self.t2 = 0.0
        self.x2 = 0.0
        self.v2 = 0.0

        self.ap = 0.0
        self.vp = 0.0

    def init(self, vs, vp, am, jm):
        self.has_uniform = jm * (vp - vs) > am * am
        self.vs = vs
        self.vp = vp
        self.jm = jm

        if self.has_uniform:
            self.ap = am

            self.t1 = am / jm
            self.t2 = (vp - vs) / am
            dt2 = self.t2 - self.t1

            self.v1 = vs + 0.5 * am * self.t1
            self.v2 = vp - 0.5 * am * self.t1

            self.x1 = vs * self.t1 + (1 / 6.0) * am * self.t1 * self.t1
            self.x2 = self.x1 + self.v1 * dt2 + 0.5 * am * dt2 * dt2

            self.total_time = self.t2 + self.t1
            self.total_distance = ((vp * vp - vs * vs) / (2.0 * am) +
                                   0.5 * (vs + vp) * self.t1)
        else:
            self.ap = math.sqrt(jm * (vp - vs))

            self.t1 = self.ap / jm
            self.t2 = self.t1

            self.v1 = vs + 0.5 * self.ap * self.ap / jm
            self.v2 = self.v1

            self.x1 = vs * self.t1 + (1 / 6.0) * self.ap * self.t1 * self.t1
            self.x2 = self.x1

            self.total_time = 2.0 * math.sqrt((vp - vs) / jm)
            self.total_distance = (vs + vp) * math.sqrt((vp - vs) / jm)

    def get_distance(self, t):
        if t <= 0:
            return 0
        if t < self.t1:
            return self.vs * t + (1 / 6.0) * self.jm * t * t * t
        if self.has_uniform and t < self.t2:
            _t = t - self.t1
            return self.x1 + self.v1 * _t + 0.5 * self.ap * _t * _t
        if t < self.total_time:
            _t = self.total_time - t
            return self.total_distance - self.vp * _t + (1 / 6.0) * self.jm * _t * _t * _t
        return self.total_distance

    def get_velocity(self, t):
        if t <= 0:
            return self.vs
        if t < self.t1:
            return self.vs + 0.5 * self.jm * t * t
        if self.has_uniform and t < self.t2:
            return self.v1 + self.ap * (t - self.t1)
        if t < self.total_time:
            _t = self.total_time - t
            return self.vp - 0.5 * self.jm * _t * _t
        return self.vp

    def get_acceleration(self, t):
        if t <= 0:
            return 0
        if t < self.t1:
            return self.jm * t
        if self.has_uniform and t < self.t2:
            return self.ap
        if t < self.total_time:
            return self.jm * (self.total_time - t)
        return 0

    def __repr__(self):
        return (f"SCurveAccel(\n"
                f"    t1: {self.t1}, t2: {self.t2}, tot_t: {self.total_time},\n"
                f"    x1: {self.x1}, x2: {self.x2}, tot_x: {self.total_distance},\n"
                f"    vs: {self.vs}, v1: {self.v1}, v2: {self.v2}, vp: {self.vp},\n"
                f"    ap: {self.ap}\n"
                f")")


class SCurve:
    S_CURVE_FAILED = 0
    S_CURVE_SUCCESS = 1

    def __init__(self):
        self.has_const = False
        self.direction = 1.0
        self.vp = 0.0
        self.vs = 0.0
        self.as_ = 0.0
        self.jm = 0.0

        self.t0 = 0.0
        self.x0 = 0.0

        self.xs = 0.0
        self.x1 = 0.0
        self.x2 = 0.0
        self.xe = 0.0

        self.process1 = SCurveAccel()
        self.ts1 = 0.0
        self.xs1 = 0.0
        self.t1 = 0.0

        self.t2 = 0.0
        self.process3 = SCurveAccel()

        self.total_time = 0.0

    def __repr__(self):
        return (f"SCurve(\n"
                f"    has_const: {self.has_const}, dir: {self.direction},\n"
                f"    vp: {self.vp},"
                f"    xs: {self.xs}, x1: {self.x1}, x2: {self.x2}, xe: {self.xe},\n"
                f"    ts1: {self.ts1}, t1: {self.t1}, t2: {self.t2}, tot: {self.total_time},\n"
                f"    proc1: {self.process1},\n"
                f"    proc3: {self.process3},\n"
                f")")

    def init(self, xs, xe, vs, as_, vm, am, jm):
        vm = abs(vm)
        am = abs(am)
        jm = abs(jm)

        dir_ = 1.0 if xe > xs else -1.0
        self.direction = dir_
        self.xs = xs
        self.xe = xe
        length = abs(xe - xs)
        vs = vs * dir_
        as_ = as_ * dir_

        self.vs = vs
        self.as_ = as_
        self.jm = jm

        if abs(vs) > vm or abs(as_) > am:
            return SCurve.S_CURVE_FAILED

        vp = vm
        vs0 = None
        dx0 = None
        vp_min = None
        vs1 = None
        ts1 = None

        if as_ < 0:
            vs0 = vs - 0.5 * as_ * as_ / jm
            if abs(vs0) > vm:
                return SCurve.S_CURVE_FAILED
            vp_min = vs0

            ts0 = -as_ / jm
            self.t0 = ts0

            dx0 = vs * ts0 + (1 / 3.0) * as_ * ts0 * ts0
            self.x0 = xs + dir_ * dx0
            vs1 = vs0
            ts1 = 0
        else:
            vp_min = vs + 0.5 * as_ * as_ / jm
            vs0 = vs
            dx0 = 0
            self.t0 = 0
            self.x0 = xs

            if vm < vp_min:
                return SCurve.S_CURVE_FAILED

            ts1 = as_ / jm
            vs1 = vs - 0.5 * as_ * ts1
        vp_min = max(0, vp_min)
        len0 = length - dx0

        self.process1.init(vs1, vp, am, jm)
        self.process3.init(0, vp, am, jm)
        self.xs1 = self.process1.get_distance(ts1)
        dx1 = self.process1.total_distance - self.xs1
        dx3 = self.process3.total_distance
        x_const = len0 - dx1 - dx3

        if x_const > 0:
            self.has_const = True
            self.ts1 = ts1
            self.t1 = self.t0 + self.process1.total_time - ts1

            t_const = x_const / vm
            self.t2 = self.t1 + t_const
            self.total_time = self.t2 + self.process3.total_time

            self.x1 = self.x0 + self.direction * dx1
            self.x2 = self.x1 + self.direction * x_const
            self.vp = vm
            return SCurve.S_CURVE_SUCCESS

        l = vp_min
        r = vm
        delta_d = len0

        cnt = 0

        while (r - vp_min) > 0.001:
            cnt += 1
            mid = 0.5 * (l + r)
            self.process1.init(vs1, mid, am, jm)
            self.process3.init(0, mid, am, jm)
            dx1 = self.process1.total_distance - self.process1.get_distance(ts1)
            dx3 = self.process3.total_distance
            delta_d = dx1 + dx3 - len0
            if -0.001 < delta_d < 0.001:
                r = l = mid
                break
            if delta_d > 0:
                r = mid
            else:
                l = mid

        print(cnt)

        if delta_d > 0.001:
            return SCurve.S_CURVE_FAILED

        self.has_const = False
        self.ts1 = ts1
        self.t1 = self.t0 + self.process1.total_time - ts1
        self.t2 = self.t1
        self.total_time = self.t2 + self.process3.total_time

        self.x1 = self.x0 + self.direction * dx1
        self.x2 = self.x1
        self.vp = r

        return SCurve.S_CURVE_SUCCESS

    def calc_x(self, t):
        if t <= 0:
            return self.xs
        if t < self.t0:
            t2 = t * t
            t3 = t2 * t
            return self.xs + self.direction * (self.vs * t +
                                               0.5 * self.as_ * t2 +
                                               (1 / 6.0) * self.jm * t3)
        if t < self.t1:
            return self.x0 + self.direction * (self.process1.get_distance(t + self.ts1 - self.t0) - self.xs1)
        if self.has_const and t < self.t2:
            return self.x1 + self.direction * self.vp * (t - self.t1)
        if t < self.total_time:
            return self.xe - self.direction * self.process3.get_distance(self.total_time - t)
        return self.xe

    def calc_v(self, t):
        if t <= 0:
            return self.direction * self.vs
        if t < self.t0:
            return self.direction * (self.vs + self.as_ * t + 0.5 * self.jm * t * t)
        if t < self.t1:
            return self.direction * self.process1.get_velocity(t + self.ts1 - self.t0)
        if self.has_const and t < self.t2:
            return self.direction * self.vp
        if t < self.total_time:
            return self.direction * self.process3.get_velocity(self.total_time - t)
        return 0.0

    def calc_a(self, t):
        if t <= 0:
            return self.direction * self.as_
        if t < self.t0:
            return self.direction * (self.as_ + self.jm * t)
        if t < self.t1:
            return self.direction * self.process1.get_acceleration(t + self.ts1 - self.t0)
        if self.has_const and t < self.t2:
            return 0.0
        if t < self.total_time:
            return -self.direction * self.process3.get_acceleration(self.total_time - t)
        return 0.0
